File: frontend/src/utils/projectStages.js
/** Varsayılan proje aşamaları (firma ayarlarından özelleştirilir). */
export const DEFAULT_PROJECT_STAGES = [
  { key: "planning", label: "Planlama", tone: "slate" },
  { key: "active", label: "Devam Ediyor", tone: "blue" },
  { key: "on_hold", label: "Beklemede", tone: "amber" },
  { key: "completed", label: "Tamamlandı", tone: "emerald", is_final: true },
];

const TONE_CLASS = {
  slate: "bg-slate-100 text-slate-600",
  blue: "bg-blue-50 text-blue-700",
  amber: "bg-amber-50 text-amber-700",
  emerald: "bg-emerald-50 text-emerald-700",
  rose: "bg-rose-50 text-rose-700",
  violet: "bg-violet-50 text-violet-700",
  indigo: "bg-indigo-50 text-indigo-700",
};

export const PROJECT_STAGE_TONES = Object.keys(TONE_CLASS);

export function stageToneClass(tone) {
  return TONE_CLASS[tone] || TONE_CLASS.slate;
}

export function slugStageKey(label, used = new Set()) {
  const base = String(label || "asama")
    .toLocaleLowerCase("tr-TR")
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/[^a-z0-9]+/g, "_")
    .replace(/^_|_$/g, "")
    .slice(0, 24) || "asama";
  let key = base;
  let i = 2;
  while (used.has(key)) {
    key = `${base}_${i}`;
    i += 1;
  }
  return key;
}

/** API / şirket kaydından gelen listeyi güvenli varsayılanlara indirger. */
export function normalizeProjectStages(raw) {
  const list = Array.isArray(raw) ? raw : [];
  const used = new Set();
  const out = [];
  for (const row of list) {
    if (!row || typeof row !== "object") continue;
    let key = String(row.key || "").trim();
    const label = String(row.label || "").trim();
    if (!label) continue;
    if (!key) key = slugStageKey(label, used);
    if (used.has(key)) continue;
    used.add(key);
    const tone = PROJECT_STAGE_TONES.includes(row.tone) ? row.tone : "slate";
    out.push({
      key,
      label,
      tone,
      is_final: !!row.is_final,
    });
  }
  if (out.length < 2) return DEFAULT_PROJECT_STAGES.map((s) => ({ ...s }));
  if (!out.some((s) => s.is_final)) {
    out[out.length - 1] = { ...out[out.length - 1], is_final: true, tone: out[out.length - 1].tone || "emerald" };
  }
  // Tek final kalsın
  let sawFinal = false;
  return out.map((s) => {
    if (!s.is_final) return s;
    if (sawFinal) return { ...s, is_final: false };
    sawFinal = true;
    return s;
  });
}

export function projectStageMap(stages) {
  const list = normalizeProjectStages(stages);
  return Object.fromEntries(list.map((s) => [s.key, [s.label, stageToneClass(s.tone)]]));
}

export function finalProjectStageKey(stages) {
  const list = normalizeProjectStages(stages);
  return (list.find((s) => s.is_final) || list[list.length - 1]).key;
}

export function projectStageLabel(stages, key) {
  const map = projectStageMap(stages);
  return map[key]?.[0] || key || "—";
}

File: backend/project_stages.py
"""Firma bazlı proje aşamaları (kart durumu seçici)."""
from __future__ import annotations

import re
import unicodedata
from typing import Any, List

DEFAULT_PROJECT_STAGES = [
    {"key": "planning", "label": "Planlama", "tone": "slate"},
    {"key": "active", "label": "Devam Ediyor", "tone": "blue"},
    {"key": "on_hold", "label": "Beklemede", "tone": "amber"},
    {"key": "completed", "label": "Tamamlandı", "tone": "emerald", "is_final": True},
]

ALLOWED_TONES = {"slate", "blue", "amber", "emerald", "rose", "violet", "indigo"}


def _slug(label: str, used: set) -> str:
    raw = unicodedata.normalize("NFKD", str(label or "asama"))
    raw = "".join(c for c in raw if not unicodedata.combining(c))
    base = re.sub(r"[^a-z0-9]+", "_", raw.lower()).strip("_")[:24] or "asama"
    key = base
    i = 2
    while key in used:
        key = f"{base}_{i}"
        i += 1
    return key


def normalize_project_stages(raw: Any) -> List[dict]:
    rows = raw if isinstance(raw, list) else []
    used: set = set()
    out: List[dict] = []
    for row in rows:
        if not isinstance(row, dict):
            continue
        label = str(row.get("label") or "").strip()
        if not label:
            continue
        key = str(row.get("key") or "").strip()
        if not key:
            key = _slug(label, used)
        if key in used:
            continue
        used.add(key)
        tone = row.get("tone") if row.get("tone") in ALLOWED_TONES else "slate"
        out.append({
            "key": key,
            "label": label[:60],
            "tone": tone,
            "is_final": bool(row.get("is_final")),
        })
    if len(out) < 2:
        return [dict(s) for s in DEFAULT_PROJECT_STAGES]
    if not any(s.get("is_final") for s in out):
        out[-1]["is_final"] = True
        out[-1]["tone"] = out[-1].get("tone") or "emerald"
    saw = False
    cleaned = []
    for s in out:
        if s.get("is_final"):
            if saw:
                s = {**s, "is_final": False}
            else:
                saw = True
        cleaned.append(s)
    return cleaned


def final_stage_key(stages: Any) -> str:
    stages = normalize_project_stages(stages)
    for s in stages:
        if s.get("is_final"):
            return s["key"]
    return stages[-1]["key"]
